### PR TITLE
Add new Apache or MIT license rule  #3738

### LINKED ISSUE #3738: Surprising false positives of mit_or_gpl-3.0_17.RULE
### Description

With variations of the copyright header in the test file, the rule mit_or_gpl-3.0_17.RULE is matched for a file that is clearly MIT and Apache 2.0. The first line of the text is not marked as matched by the rule, but if not present will prevent the false positive.

### How To Reproduce

Create the test file test.py with the following content:

```
# Copyright (c) Someone and affiliates.
#
# This source code is licensed under both the MIT license found in the
# LICENSE_MIT file in the root directory of this source tree and the Apache
# License, Version 2.0 found in the LICENSE_APACHE file in the root directory
# of this source tree.
```

Run any scancode 32, including the latest one, with the following flags:
 -clpu --only-findings -n35 --license-text --verbose --json results.json ./test.py

The output, formatted, will look more or less (I've removed my system information) like this:
```
{
 "headers": [
   {
     "tool_name": "scancode-toolkit",
     "tool_version": "32.0.6",
     "options": {
       "input": [
         "/test_sc.py"
       ],
       "--copyright": true,
       "--json": "test.json",
       "--license": true,
       "--license-text": true,
       "--only-findings": true,
       "--package": true,
       "--processes": "35",
       "--url": true,
       "--verbose": true
     },
     "notice": "Generated with ScanCode and provided on an \"AS IS\" BASIS, WITHOUT WARRANTIES\nOR CONDITIONS OF ANY KIND, either express or implied. No content created from\nScanCode should be considered or used as legal advice. Consult an Attorney\nfor any legal advice.\nScanCode is a free software code scanning tool from nexB Inc. and others.\nVisit https://github.com/nexB/scancode-toolkit/ for support and download.",
     "start_timestamp": "2024-04-12T093906.516396",
     "end_timestamp": "2024-04-12T093910.030508",
     "output_format_version": "3.0.0",
     "duration": 3.5141329765319824,
     "message": null,
     "errors": [],
     "warnings": [],
     "extra_data": {
       "spdx_license_list_version": "3.21",
       "files_count": 1
     }
   }
 ],
 "packages": [],
 "dependencies": [],
 "license_detections": [
   {
     "identifier": "mit_or_gpl_3_0__and_apache_2_0-c8dec2e8-7c20-200b-983a-ad0cf4c9dbbb",
     "license_expression": "(mit OR gpl-3.0) AND apache-2.0",
     "detection_count": 1
   }
 ],
 "files": [
   {
     "path": "test_sc.py",
     "type": "file",
     "package_data": [],
     "for_packages": [],
     "detected_license_expression": "(mit OR gpl-3.0) AND apache-2.0",
     "detected_license_expression_spdx": "(MIT OR GPL-3.0-only) AND Apache-2.0",
     "license_detections": [
       {
         "license_expression": "(mit OR gpl-3.0) AND apache-2.0",
         "matches": [
           {
             "score": 60.87,
             "start_line": 3,
             "end_line": 4,
             "matched_length": 14,
             "match_coverage": 60.87,
             "matcher": "3-seq",
             "license_expression": "mit OR gpl-3.0",
             "rule_identifier": "mit_or_gpl-3.0_17.RULE",
             "rule_relevance": 100,
             "rule_url": ["https://github.com/nexB/scancode-toolkit/tree/develop/src/licensedcode/data/rules/mit_or_gpl-3.0_17.RULE"](https://github.com/nexB/scancode-toolkit/tree/develop/src/licensedcode/data/rules/mit_or_gpl-3.0_17.RULE),
             "matched_text": "# This source code is licensed under both the MIT license found in the\n# LICENSE_MIT file in the root directory of this source tree and the Apache"
           },
           {
             "score": 100,
             "start_line": 4,
             "end_line": 5,
             "matched_length": 6,
             "match_coverage": 100,
             "matcher": "2-aho",
             "license_expression": "apache-2.0",
             "rule_identifier": "apache-2.0_182.RULE",
             "rule_relevance": 100,
             "rule_url": ["https://github.com/nexB/scancode-toolkit/tree/develop/src/licensedcode/data/rules/apache-2.0_182.RULE"](https://github.com/nexB/scancode-toolkit/tree/develop/src/licensedcode/data/rules/apache-2.0_182.RULE),
             "matched_text": "# LICENSE_MIT file in the root directory of this source tree and the Apache\n# License, Version 2.0 found in the LICENSE_APACHE file in the root directory"
           },
           {
             "score": 90,
             "start_line": 5,
             "end_line": 5,
             "matched_length": 2,
             "match_coverage": 100,
             "matcher": "2-aho",
             "license_expression": "apache-2.0",
             "rule_identifier": "apache-2.0_161.RULE",
             "rule_relevance": 90,
             "rule_url": ["https://github.com/nexB/scancode-toolkit/tree/develop/src/licensedcode/data/rules/apache-2.0_161.RULE"](https://github.com/nexB/scancode-toolkit/tree/develop/src/licensedcode/data/rules/apache-2.0_161.RULE),
             "matched_text": "# License, Version 2.0 found in the LICENSE_APACHE file in the root directory"
           }
         ],
         "identifier": "mit_or_gpl_3_0__and_apache_2_0-c8dec2e8-7c20-200b-983a-ad0cf4c9dbbb"
       }
     ],
     "license_clues": [],
     "percentage_of_license_text": 44,
     "copyrights": [
       {
         "copyright": "Copyright (c) Someone and affiliates",
         "start_line": 1,
         "end_line": 1
       }
     ],
     "holders": [
       {
         "holder": "Someone and affiliates",
         "start_line": 1,
         "end_line": 1
       }
     ],
     "authors": [],
     "urls": [],
     "scan_errors": []
   }
 ]
}
```

### System configuration

> For bug reports, it really helps us to know:

* What OS are you running on? Linux
* What version of scancode-toolkit was used to generate the scan file? 32.0.6 (same results when trying with 32.1.0)
* What installation method was used to install/run scancode? source download

### PR DESCRIPTION
**Pull request by vasily-pozdnyakov**
Fixes #3738

### Tasks

* [x] Reviewed [contribution guidelines](https://github.com/nexB/scancode-toolkit/blob/develop/CONTRIBUTING.rst)
* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [x] Commits are in uniquely-named feature branch and has no merge conflicts 📁

### CODE REVIEW

**Comment by vasily-pozdnyakov** [comment]
Hi, I am new to new rules creation and have a couple of questions:
1. What does scancode-reindex-licenses do? It does not introduce any visible changes (validation is clear, what about reindex?).
2. Why most of the rules are weak (without "{{}}")? If I understand correctly, it might bring a lot of false positives (let's say, there is a matching license notice for MIT, but with a different license name (Apache) - the scancode might detect it incorrectly), is it like that?

**Comment by AyanSinhaMahapatra** [comment]
@vasily-pozdnyakov welcome and thanks for the PR!

> What does scancode-reindex-licenses do? It does not introduce any visible changes (validation is clear, what about reindex?).

When you get SCTK from a git checkout and install scancode to run locally (or get scancode from github releases or pip install) you have the license indexes at `scancode-toolkit/src/licensedcode/data/cache/license_index/` which contains the license index (pre-built if you're downloading the release/via pip) that is used for license detection when you run scancode. Now if you are updating/adding new rule/licenses in scancode you want to run `scancode-reindex-licenses` so next time you run a scan locally these rules are in the index and are used in license detection, but this does no changes in the repository of rules.

> Why most of the rules are weak (without "{{}}")? If I understand correctly, it might bring a lot of false positives (let's say, there is a matching license notice for MIT, but with a different license name (Apache) - the scancode might detect it incorrectly), is it like that?


That's mostly correct, this could (and does also) generate a lot of false positives in a lot of cases. See https://github.com/nexB/scancode-toolkit/issues/2878

We are working on resolving this and I have a pending PR at https://github.com/nexB/scancode-toolkit/pull/3254 testing out adding required phrases automatically and massively across all the rules from the following:
1. license names and other license keywords
2. from other instances of these required phrases added in other rules.
This would improve the accuracy significantly for these rules and also make sure when we add a required phrase, same is marked across the same phrases present in all rules. There is just a bit of work remaining on testing this and verifying this works correctly across all the rules. But this addition of required phrases would be continuous, or a massive one-time effort otherwise.

File: src/licensedcode/data/rules/mit_and_apache-2.0_10.RULE [hunk 1]
```diff
@@ -0,0 +1,10 @@
+---
+license_expression: mit AND apache-2.0
+is_license_notice: yes
+relevance: 100
+---
```
Comment (AyanSinhaMahapatra) [inline src/licensedcode/data/rules/mit_and_apache-2.0_10.RULE:8]:
```suggestion
referenced_filenames:
   - LICENSE_MIT
   - LICENSE_APACHE
---
```

File: src/licensedcode/data/rules/mit_and_apache-2.0_10.RULE [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+license_expression: mit AND apache-2.0
```
Comment (pombredanne) [inline src/licensedcode/data/rules/mit_and_apache-2.0_10.RULE]:
```suggestion
license_expression: mit OR apache-2.0
```

As weird as it sounds this is choice of license, not a conjunction. This is the license of BUCK afaik at https://github.com/facebook/buck2 ans https://github.com/facebook/buck2/blob/732db19980a934de534193ec2d4e330a1d251575/ci.bzl#L3 and this a choice of licenses.

**Review by pombredanne — CHANGES_REQUESTED** [review]
Thanks. Just a minor but important nit wrt. to the license expression. The norm for MIT and Pache in the Rust ecosystem is to be a choice and buck2 and this new rules is not an exception.

We could also later add more rules for derived notices such as https://github.com/facebook/buck2?tab=readme-ov-file#license

**Comment by AyanSinhaMahapatra** [comment]
Thanks++ @vasily-pozdnyakov, merging!

**Pull request merged**